keylime/da/attest.py

```diff
@@ -0,0 +1,154 @@
+#!/usr/bin/python3
+import argparse
+from datetime import datetime
+
+from keylime import cloud_verifier_common, config, keylime_logging
+from keylime.da import record
+
+logger = keylime_logging.init_logging("durable_attestation_fetch_and_replay")
+
+
+def main() -> None:
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "-u",
+        "--uuid",
+        action="store",
+        dest="agent_uuid",
+        default="all",
+        help='UUID for the agent to attest (default "all")',
+    )
+    parser.add_argument(
+        "-s",
+        "--start",
+        action="store",
+        dest="start_date",
+        default="first",
+        help='start date for attestation window in either IS0 8601 format (e.g., "2008-09-03T20:56:35"), or keyworkd "first"/"none"',
+    )
```

Comment:
Attestation has to start at the very beginning of a machine's log. I am not sure how one could enter some arbitrary time here. You cannot miss out on the first 100 measurements and just start with the 101th that occurred after 12:00 am.

Reply:
You have to know the state of the PCR 10 after 100 measurements to be able to resume at the 101th. Do you know the state at any point in time?

Reply:
If I understand DA correctly where I assume every IMA measurement log comes in separately when it occurred, you would have to search backwards to the beginning of the log of a machine at 12:00am, so let's say 11:50am to get the first measurement log entry for IMA (typically boot_aggregate) and then you can start replaying the other ones on top of it. For everything else you would otherwise have to know the state of the PCRs (or IMA PCR10) at any point in time so you could work with an arbitrary 'window'.

Reply:
(Ok,  back from vacation). You're correct, for **IMA only** we need to replay the whole data series of attestation artifacts in order to get the correct one. I am most certainly not going to modify the core `verifier` code at this point (the most relevant aspect of these PRs is to be able run **unmodified** attestation code), but plan to address this particular case on a subsequent PR (either on the code or at least via documentation).

Reply:
The support for 'all' also seems to be complicated where you would have to search backwards for each individual machine that is part of the entire attestation window...

Reply:
Computationally intensive for sure, but feasible.

Reply:
This may require indexed data if one wanted to find the boot_aggregate entry from 200 days ago ... iff one wanted to validate quotes.

Reply:
If one replays the whole series from "beginning of time" the `boot_aggregate` will be there, given that it is part of the "attestation artifacts" recorded on the datastore. However, the plugin architecture allows one to create indexes on a backend on `keylime/da/examples` if one decides this is crucial. Please note that these `examples` are **not** officially supported for the time being (it is a subject of active discussion at this point), and all that is outside the scope of this particular PR.

Reply:
"beginning of time": That's true. 

Measured boot relative to durable attestation is trivial since the rather small log is transferred every time.
IMA attestation relative to durable attestation is non-trivial primarily due to the statefulness of IMA attestation support we have in Keylime that we added to reduce the unnecessary transfer of the entire IMA log every few seconds when attestation occurs. Instead we only transfer the new log entries now and extend known PCR 10 state in an in-memory PCR. For support of IMA attestation we'll need efficient access to data, to find the beginning of the log, as well as being able to detect when a system was rebooted or attestation was restarted so we can reset PCR 10 (or all of them). The beginnings of IMA attestation logs can presumably be detected when the boot_aggregate entry shows up.

Once you have (delta) IMA log entries in the durable attestation replay you won't be able to do the easier measured boot replay. One would have to split the quote into two quotes to be able to separated them and only focus on measured boot replay. One quote would be to be for measured boot and the other for IMA attestation...

Reply:
Agreed. We can file feature enhancements to make this feature cover more scenarios, but that shouldn't keep it from being merged while it is useful now.